File: src/common/winq/statement/StatementCreateTrigger.cpp
//
// Created by sanhuazhang on 2019/05/02
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "WINQ.h"

namespace WCDB {

StatementCreateTrigger::StatementCreateTrigger() = default;

StatementCreateTrigger::~StatementCreateTrigger() = default;

StatementCreateTrigger& StatementCreateTrigger::createTrigger(const UnsafeStringView& trigger)
{
    syntax().trigger = trigger;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::schema(const Schema& schema)
{
    syntax().schema = schema;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::temp()
{
    syntax().temp = true;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::ifNotExists()
{
    syntax().ifNotExists = true;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::before()
{
    syntax().specifiedTiming = true;
    syntax().timing = SyntaxType::Timing::Before;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::after()
{
    syntax().specifiedTiming = true;
    syntax().timing = SyntaxType::Timing::After;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::insteadOf()
{
    syntax().specifiedTiming = true;
    syntax().timing = SyntaxType::Timing::InsteadOf;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::delete_()
{
    syntax().event = SyntaxType::Event::Delete;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::insert()
{
    syntax().event = SyntaxType::Event::Insert;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::update()
{
    syntax().event = SyntaxType::Event::Update;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::column(const Column& column)
{
    syntax().columns.push_back(column);
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::on(const UnsafeStringView& table)
{
    syntax().table = table;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::forEachRow()
{
    syntax().forEachFow = true;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::when(const Expression& condition)
{
    syntax().condition = condition;
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::execute(const StatementInsert& statement)
{
    syntax().stmts.push_back(SyntaxType::STMT::Insert);
    syntax().inserts.push_back(statement);
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::execute(const StatementUpdate& statement)
{
    syntax().stmts.push_back(SyntaxType::STMT::Update);
    syntax().updates.push_back(statement);
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::execute(const StatementDelete& statement)
{
    syntax().stmts.push_back(SyntaxType::STMT::Delete);
    syntax().deletes.push_back(statement);
    return *this;
}

StatementCreateTrigger& StatementCreateTrigger::execute(const StatementSelect& statement)
{
    syntax().stmts.push_back(SyntaxType::STMT::Select);
    syntax().selects.push_back(statement);
    return *this;
}

} // namespace WCDB

File: src/swift/winq/statement/StatementCreateTrigger.swift
//
// Created by qiuwenchen on 2022/7/14.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import WCDB_Private

public final class StatementCreateTrigger: Identifier<CPPStatementCreateTrigger>, Statement {
    public init() {
        super.init(with: WCDBStatementCreateTriggerCreate())
    }

    @discardableResult
    public func create(trigger: String, isTemp: Bool = false) -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigTrigger(cppObj, trigger.cString)
        if  isTemp {
            WCDBStatementCreateTriggerConfigTemp(cppObj)
        }
        return self
    }

    @discardableResult
    public func of(schema schemaConvertible: SchemaConvertible) -> StatementCreateTrigger {
        let schema = schemaConvertible.asSchema()
        withExtendedLifetime(schema) {
            WCDBStatementCreateTriggerConfigSchema(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func ifNotExists() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigIfNotExist(cppObj)
        return self
    }

    @discardableResult
    public func before() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigBefore(cppObj)
        return self
    }

    @discardableResult
    public func after() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigAfter(cppObj)
        return self
    }

    @discardableResult
    public func insteadOf() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigInsteadOf(cppObj)
        return self
    }

    @discardableResult
    public func delete() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigDelete(cppObj)
        return self
    }

    @discardableResult
    public func insert() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigInsert(cppObj)
        return self
    }

    @discardableResult
    public func update() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigUpdate(cppObj)
        return self
    }

    @discardableResult
    public func of(columns columnConvertibleList: ColumnConvertible...) -> StatementCreateTrigger {
        of(columns: columnConvertibleList)
    }

    @discardableResult
    public func of(columns columnConvertibleList: [ColumnConvertible]) -> StatementCreateTrigger {
        let columns = columnConvertibleList.map {$0.asColumn()}
        withExtendedLifetime(columns) {
            let cppColumns = columns.map {$0.cppObj}
            cppColumns.withUnsafeBufferPointer({
                WCDBStatementCreateTriggerConfigColumns(cppObj, $0.baseAddress, Int32($0.count))
            })
        }
        return self
    }

    @discardableResult
    public func on(table: String) -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigTable(cppObj, table.cString)
        return self
    }

    @discardableResult
    public func forEachRow() -> StatementCreateTrigger {
        WCDBStatementCreateTriggerConfigForEachRow(cppObj)
        return self
    }

    @discardableResult
    public func when(_ expressionConvertible: ExpressionConvertible) -> StatementCreateTrigger {
        let expression = expressionConvertible.asExpression()
        withExtendedLifetime(expression) {
            WCDBStatementCreateTriggerConfigWhen(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func execute(_ insert: StatementInsert) -> StatementCreateTrigger {
        withExtendedLifetime(insert) {
            WCDBStatementCreateTriggerExecuteInsert(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func execute(_ update: StatementUpdate) -> StatementCreateTrigger {
        withExtendedLifetime(update) {
            WCDBStatementCreateTriggerExecuteUpdate(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func execute(_ delete: StatementDelete) -> StatementCreateTrigger {
        withExtendedLifetime(delete) {
            WCDBStatementCreateTriggerExecuteDelete(cppObj, $0.cppObj)
        }
        return self
    }

    @discardableResult
    public func execute(_ select: StatementSelect) -> StatementCreateTrigger {
        withExtendedLifetime(select) {
            WCDBStatementCreateTriggerExecuteSelect(cppObj, $0.cppObj)
        }
        return self
    }

}
